Tensorify compute on Python scalars

bobrenjc93 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #137612
* __->__ #136674

Signed-off-by: Bob Ren <bobrenfb.com>

Comandeered from https://github.com/pytorch/pytorch/pull/130228 as I'm helping @ezyang w/ shipping dynamic float arguments in PT2. This starts with supporting torch.ops.aten.mul. I'll stack on top support for other operators in subsequent PRs to keep this scoped to the mechanics of the fx pass.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @rec

bobrenjc93 committed:
Tensorify compute on Python scalars

Signed-off-by: Bob Ren <bobrenfb.com>

[ghstack-poisoned]

ezyang commented on test/dynamo/test_misc.py:
```diff
@@ -2644,6 +2644,19 @@ def fn(x):
+        x = torch.arange(3)
+        self.assertEqual(fn(x, 2.0), fn_opt(x, 2.0))
+        self.assertEqual(fn(x, 3.0), fn_opt(x, 3.0))
+        self.assertEqual(cnt.frame_count, 1)
```
It's also possible Inductor "accidentally" supported passing a SymFloat into an addition kernel, so it might be good to peek at the IR post pass and assert that it looks like what we accept.

ezyang replied:
Also, this is better in the test/inductor/test_torchinductor_dynamic_shapes.py or something, since you're going all the way to inductor here

ezyang replied:
Meh, I can't think of a good way to get at the post graph transform graph, except maybe something like was done in https://github.com/pytorch/pytorch/pull/136508

ezyang commented on torch/fx/passes/tensorify_python_scalars.py:
```diff
@@ -0,0 +1,245 @@
+                        is torch.ops.aten._local_scalar_dense.default
+                    ):
+                        # TODO: dtype conversion, so that we don't keep at too
+                        # low precision
```
So, specifically, what this means is if someone passes in a f32[] scalar tensor and then immediately calls item() on it, the logic here would just take the f32[] tensor directly, which means we'd do all the subsequent tensor compute in f32[] rather than in f64[] (which is what the Python compute would have done). It's worth fixing this, even if it's as simple as preventing us from assigning tensor/sym proxies when the original tensor is f32 (this will never happen for SymFloat to tensor argument changes).

ezyang commented on torch/fx/passes/tensorify_python_scalars.py:
```diff
@@ -0,0 +1,245 @@
+                        ):
+                            return go(
+                                graph.call_method(keypath[0].name, (node,)), keypath[1:]
+                            )
```
I uh... I think this doesn't do anything anymore?

ezyang commented on torch/fx/passes/tensorify_python_scalars.py:
```diff
@@ -0,0 +1,245 @@
+                                    node.meta["val"].dtype,
+                                ),
+                            )
+                        except NotImplementedError:
```
can you make the scope of this try-catch smaller? It's just for the `_sympy_interp` right?

ezyang commented on torch/fx/passes/tensorify_python_scalars.py:
```diff
@@ -0,0 +1,245 @@
+                                (
+                                    _sympy_interp(zf.node.expr).node,
+                                    node.meta["val"].dtype,
+                                ),
```
Unfortunately, this is not right for low precision dtype, because the computation dtype can differ from result dtype (this is not really your fault, because I quoted you the Numpy docs but Numpy doesn't have low prec dtypes lol). Quoting `elementwise_dtypes` doc:

```
    The DEFAULT type promotion kind computes per above, and then uses the result dtype to pick a computation
    dtype by mapping low precision floating point and complex dtypes as follows:

      float16   -> float32
      bfloat16  -> float32
      complex32 -> complex64
```

So your logic here would incorrectly downconvert the scalar to float16 before adding it to a float16 tensor, which would lose precision even though the internal compute is done in float32.

This part will be good to test carefully.

ezyang commented:
Nice work fixing the tests! After some slight refinements, the next step is scaleout!

bobrenjc93 committed:
Update on "Tensorify compute on Python scalars"


Signed-off-by: Bob Ren <bobrenfb.com>

Comandeered from https://github.com/pytorch/pytorch/pull/130228 as I'm helping ezyang w/ shipping dynamic float arguments in PT2. This starts with supporting torch.ops.aten.mul. I'll stack on top support for other operators in subsequent PRs to keep this scoped to the mechanics of the fx pass.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Tensorify compute on Python scalars"


Signed-off-by: Bob Ren <bobrenfb.com>

Comandeered from https://github.com/pytorch/pytorch/pull/130228 as I'm helping ezyang w/ shipping dynamic float arguments in PT2. This starts with supporting torch.ops.aten.mul. I'll stack on top support for other operators in subsequent PRs to keep this scoped to the mechanics of the fx pass.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Tensorify compute on Python scalars"


Signed-off-by: Bob Ren <bobrenfb.com>

Comandeered from https://github.com/pytorch/pytorch/pull/130228 as I'm helping ezyang w/ shipping dynamic float arguments in PT2. This starts with supporting torch.ops.aten.mul. I'll stack on top support for other operators in subsequent PRs to keep this scoped to the mechanics of the fx pass.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Tensorify compute on Python scalars"


Signed-off-by: Bob Ren <bobrenfb.com>

Comandeered from https://github.com/pytorch/pytorch/pull/130228 as I'm helping ezyang w/ shipping dynamic float arguments in PT2. This starts with supporting torch.ops.aten.mul. I'll stack on top support for other operators in subsequent PRs to keep this scoped to the mechanics of the fx pass.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

bobrenjc93 commented on torch/utils/_sympy/reference.py:
```diff
@@ -421,7 +421,7 @@ def int_truediv(a, b):
     @staticmethod
     def floordiv(a, b):
-        return torch.ops.aten.floor_divide(a, b)
+        return torch.ops.aten.div.Tensor_mode(a, b, rounding_mode = 'floor')
```
I couldn't get `return torch.ops.aten.floor_divide.default(a, b)` working for the `python test/inductor/test_torchinductor_dynamic_shapes.py TestInductorDynamicCUDA.test_interpolate_ceil_eq_cuda` test.

It would output

```
MissingOperatorWithDecomp: missing decomposition
  target: aten.floor_divide.default
  args[0]: TensorBox(StorageBox(
    Pointwise(
      'cpu',
      torch.float32,
      def inner_fn(index):
          tmp0 = ops.constant(-1, torch.int64)
          tmp1 = ops.to_dtype(tmp0, torch.float32, src_dtype=torch.int64)
          tmp2 = ops.index_expr(s1, torch.float32)
          tmp3 = tmp1 + tmp2
          return tmp3
      ,
      ranges=[],
      origin_node=add_tensor,
      origins=OrderedSet([add_tensor, scalar_tensor_default_3, full...
    )
  ))
  args[1]: TensorBox(StorageBox(
    Pointwise(
      'cpu',
      torch.int64,
      def inner_fn(index):
          tmp0 = ops.constant(8, torch.int64)
          return tmp0
      ,
      ranges=[],
      origin_node=full_default_3,
      origins=OrderedSet([full_default_3])
    )
  ))

There is a decomposition available for aten.floor_divide.default in
torch._decomp.get_decompositions().  Please add this operator to the
`decompositions` list in torch._inductor.decomposition
```

But adding `aten.floor_divide.default` to the list didn't fix it. If reviewers have any suggestions on how to get `torch.ops.aten.floor_divide` to work properly, I'm happy to make the change.

ezyang replied:
No, your change here looks right, thanks

bobrenjc93 committed:
Update on "Tensorify compute on Python scalars"


Signed-off-by: Bob Ren <bobrenfb.com>

Comandeered from https://github.com/pytorch/pytorch/pull/130228 as I'm helping ezyang w/ shipping dynamic float arguments in PT2. This starts with supporting torch.ops.aten.mul. I'll stack on top support for other operators in subsequent PRs to keep this scoped to the mechanics of the fx pass.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

ezyang commented on torch/fx/passes/tensorify_python_scalars.py:
```diff
@@ -0,0 +1,245 @@
+                        # https://github.com/pytorch/pytorch/blob/main/torch/_prims_common/__init__.py#L1379
+                        compute_dtype = (
+                            torch.float32
+                            if node.meta["val"].dtype == torch.float16
```
You can't only do this for float16, this is needed also for bfloat16, among others. Check elementwise_dtypes to see if there's a way you can do it without duplicating all the logic here.

bobrenjc93 committed:
Update on "Tensorify compute on Python scalars"


Signed-off-by: Bob Ren <bobrenfb.com>

Comandeered from https://github.com/pytorch/pytorch/pull/130228 as I'm helping ezyang w/ shipping dynamic float arguments in PT2. This starts with supporting torch.ops.aten.mul. I'll stack on top support for other operators in subsequent PRs to keep this scoped to the mechanics of the fx pass.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

bobrenjc93 commented on torch/fx/passes/tensorify_python_scalars.py:
```diff
@@ -0,0 +1,241 @@
+                                    compute_dtype,
+                                ),
+                            )
+                            res.meta = node.meta
```
This passes all the tests but feels weird to do. Ideally I'd do something like

```
res.meta[
    "val"
] = torch.ops.prims.convert_element_type.default(
    a.meta["val"], compute_dtype
)
```

but we don't always have res.meta["val"].  For example in `python test/inductor/test_torchinductor.py CpuTests.test_pow_symfloat_cpu` we don't. I'm still digesting metas/vals/fake tensors/etc. so if any of the reviewers have a better way to populate meta here, I'm all ears!

ezyang commented on torch/fx/passes/tensorify_python_scalars.py:
```diff
@@ -0,0 +1,241 @@
+                compute_dtype, result_dtype = elementwise_dtypes(
+                    node.meta["val"],
+                    type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT,
+                )
```
Because you're only consulting the result dtype, don't bother with `elementwise_dtypes`, just directly call `get_computation_dtype` on the node.meta["val"].dtype

ezyang commented on torch/fx/passes/tensorify_python_scalars.py:
```diff
@@ -0,0 +1,241 @@
+            node = graph.call_function(
+                torch.ops.aten.scalar_tensor.default, (c,), {"dtype": dtype}
+            )
+            node.meta["val"] = torch.tensor(c, dtype=dtype)
```
These have to be fake tensors?

ezyang commented on torch/fx/passes/tensorify_python_scalars.py:
```diff
@@ -0,0 +1,241 @@
+
+    graph_code_log.debug(
+        "%s", lazy_format_graph_code("tensorify_python_scalars", gm, colored=True)
+    )
```
Let's do FakeTensorUpdater instead of the manual prop

ezyang commented on torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py:
```diff
@@ -177,6 +179,10 @@ def aot_dispatch_base(
+            fake_mode = detect_fake_mode()
+            if fake_mode is not None:
+                assert isinstance(fw_module, GraphModule)
+                tensorify_python_scalars(fw_module, fake_mode.shape_env)
```
Consider feature flagging this

ezyang replied:
This might be the wrong spot, it needs to be on the joint graph

ezyang replied:
You need to add another call at the joint graph

bobrenjc93 committed:
Update on "Tensorify compute on Python scalars"


Signed-off-by: Bob Ren <bobrenfb.com>

Comandeered from https://github.com/pytorch/pytorch/pull/130228 as I'm helping ezyang w/ shipping dynamic float arguments in PT2. This starts with supporting torch.ops.aten.mul. I'll stack on top support for other operators in subsequent PRs to keep this scoped to the mechanics of the fx pass.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Tensorify compute on Python scalars"


Signed-off-by: Bob Ren <bobrenfb.com>

Comandeered from https://github.com/pytorch/pytorch/pull/130228 as I'm helping ezyang w/ shipping dynamic float arguments in PT2. This starts with supporting torch.ops.aten.mul. I'll stack on top support for other operators in subsequent PRs to keep this scoped to the mechanics of the fx pass.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Tensorify compute on Python scalars"


Signed-off-by: Bob Ren <bobrenfb.com>

Comandeered from https://github.com/pytorch/pytorch/pull/130228 as I'm helping ezyang w/ shipping dynamic float arguments in PT2. This starts with supporting torch.ops.aten.mul. I'll stack on top support for other operators in subsequent PRs to keep this scoped to the mechanics of the fx pass.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

albanD commented on docs/source/conf.py:
```diff
@@ -731,6 +731,8 @@
     "is_bool_expr",
     "is_dim",
+    # torch.fx.passes.tensorify_python_scalars
+    "tensorify_python_scalars",
```
Adding new entries here is no ok.
Please make sure that your function is properly documented or made private if you don't want it visible to the end user.

In this case, you are most likely missing the appropriate .rst entry to show this function on the docs website.

bobrenjc93 committed:
Update on "Tensorify compute on Python scalars"


Signed-off-by: Bob Ren <bobrenfb.com>

Comandeered from https://github.com/pytorch/pytorch/pull/130228 as I'm helping ezyang w/ shipping dynamic float arguments in PT2. This starts with supporting torch.ops.aten.mul. I'll stack on top support for other operators in subsequent PRs to keep this scoped to the mechanics of the fx pass.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Tensorify compute on Python scalars"


Signed-off-by: Bob Ren <bobrenfb.com>

Comandeered from https://github.com/pytorch/pytorch/pull/130228 as I'm helping ezyang w/ shipping dynamic float arguments in PT2. This starts with supporting torch.ops.aten.mul. I'll stack on top support for other operators in subsequent PRs to keep this scoped to the mechanics of the fx pass.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Tensorify compute on Python scalars"


Signed-off-by: Bob Ren <bobrenfb.com>

Comandeered from https://github.com/pytorch/pytorch/pull/130228 as I'm helping ezyang w/ shipping dynamic float arguments in PT2. This starts with supporting torch.ops.aten.mul. I'll stack on top support for other operators in subsequent PRs to keep this scoped to the mechanics of the fx pass.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Tensorify compute on Python scalars"


Signed-off-by: Bob Ren <bobrenfb.com>

Comandeered from https://github.com/pytorch/pytorch/pull/130228 as I'm helping ezyang w/ shipping dynamic float arguments in PT2. This starts with supporting torch.ops.aten.mul. I'll stack on top support for other operators in subsequent PRs to keep this scoped to the mechanics of the fx pass.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

bobrenjc93 commented on torch/fx/passes/tensorify_python_scalars.py:
```diff
@@ -0,0 +1,262 @@
+            node = graph.call_function(
+                torch.ops.aten.scalar_tensor.default, (c,), {"dtype": dtype}
+            )
+            node.meta["val"] = torch.ops.aten.scalar_tensor.default(c, dtype=dtype)
```
I wonder if it makes sense to have a propogate_meta optional arg on graph.call_function

bobrenjc93 replied:
After discussion w/ @Chillee and @zou3519, I'll likely instead refactor this (outside of this diff) to look like

```
graph.insert_traced_subgraph(
    torch.ops.aten.scalar_tensor.default, (c,), {"dtype": dtype}
)	
```

Where under the hood `insert_traced_subgraph` will call `make_fx`. This has the nicety that

1) Decompositions are handled out of the box
2) We also get meta prop out of the box

bobrenjc93 committed:
Update on "Tensorify compute on Python scalars"


Signed-off-by: Bob Ren <bobrenfb.com>

Comandeered from https://github.com/pytorch/pytorch/pull/130228 as I'm helping ezyang w/ shipping dynamic float arguments in PT2. This starts with supporting torch.ops.aten.mul. I'll stack on top support for other operators in subsequent PRs to keep this scoped to the mechanics of the fx pass.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

ezyang commented on torch/_functorch/config.py:
```diff
@@ -193,6 +193,10 @@
 
+# Fx pass used to tensorify python scalars. Part of the rollout for supporting
+# dynamic float arguments in PT2.
+tensorify_python_scalars = True
```
err, were you going to killswitch this?

bobrenjc93 replied:
Sorry I think I misunderstood how JK's work within pytorch. I thought they would override these config flags by name. I put up https://github.com/pytorch/pytorch/pull/137490/files and will double check the integration w/ Colin before folding it back into this diff.

bdhirsh commented on torch/fx/passes/_tensorify_python_scalars.py:
```diff
@@ -0,0 +1,262 @@
+# our partitioning decisions (in particular, we want to avoid having to
+# be able to save floats across the partition, and passes that change what
+# device compute happen on need to happen before partitioning, but after this
+# pass). Note that some transformations have to happen before this in Dynamo,
```
just for my understanding, which transformations do we need to do earlier in dynamo to avoid specialization? (vs. the the transformation here that we are ok with delaying until we get the joint graph)

bobrenjc93 replied:
Dynamo will convert float inputs into 0d CPU tensor inputs, which then immediately have item() calls to transform them into regular float for downstream use: https://github.com/pytorch/pytorch/pull/125325

bdhirsh replied:
ah right :) thanks

bobrenjc93 committed:
Update on "Tensorify compute on Python scalars"


Signed-off-by: Bob Ren <bobrenfb.com>

Comandeered from https://github.com/pytorch/pytorch/pull/130228 as I'm helping ezyang w/ shipping dynamic float arguments in PT2. This starts with supporting torch.ops.aten.mul. I'll stack on top support for other operators in subsequent PRs to keep this scoped to the mechanics of the fx pass.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

laithsakka commented on test/dynamo/test_misc.py:
```diff
@@ -7631,6 +7631,58 @@ def my_dyn_fn(x):
+        y = 1.00048828125
+
+        self.assertEqual(fn_opt(x, y), fn(x, y))
+        self.assertEqual(cnt.frame_count, 1)
```
curious what are we trying to test here we only call fn_opt() , so its not recompilation that we are looking at?
are you tryin to assert that there is no graph breaks? if so maybe pass fullgraph= True?
or is it something related to forward and backward?

laithsakka commented on test/dynamo/test_misc.py:
```diff
@@ -7631,6 +7631,58 @@ def my_dyn_fn(x):
+            return x * y
+
+        cnt = CompileCounterWithBackend("aot_eager")
+        fn_opt = torch._dynamo.optimize(cnt)(fn)
```
nit: torch.compile(fn, backend=cnt)

laithsakka commented on test/dynamo/test_misc.py:
```diff
@@ -7631,6 +7631,58 @@ def my_dyn_fn(x):
         torch._dynamo.optimize("eager")(my_dyn_fn)(y)
 
+    @torch._dynamo.config.patch(specialize_float=False, capture_scalar_outputs=True)
+    def test_unspecialized_float_multiply_precision_bf16(self):
```
nit: It might be good to move all dynamic related tests to their own file, maybe  dynamo/test_dynamic.py

laithsakka commented on test/dynamo/test_misc.py:
```diff
@@ -7631,6 +7631,58 @@ def my_dyn_fn(x):
+        self.assertEqual(fn_opt(x, y), fn(x, y))
+        self.assertEqual(cnt.frame_count, 1)
+
+    @torch._dynamo.config.patch(specialize_float=False, capture_scalar_outputs=True)
```
all those function are the same the only difference is dtype, maybe we can create a helper function that takes dtype and call it from those?

laithsakka commented on torch/fx/passes/_tensorify_python_scalars.py:
```diff
@@ -0,0 +1,262 @@
+log = logging.getLogger(__name__)
+graph_code_log = torch._logging.getArtifactLogger(__name__, "graph_code")
+
+# The general shape of this transformation is to look for Tensor operations
```
I think we should add an instruction count microbenchmark for this case, to be able to understand the effect of this conversion on compile time.

bobrenjc93 replied:
This is a bit difficult since without this pass, we can't even finish compiling. Inductor will choke on the .item() calls.

laithsakka replied:
I meant a micro benchmark where we compile a function with float inputs. does this PR will always convert all float input to tensors?

bobrenjc93 replied:
This one doesn't but the following PR does: https://github.com/pytorch/pytorch/pull/125325

laithsakka commented on torch/fx/passes/_tensorify_python_scalars.py:
```diff
@@ -0,0 +1,262 @@
+# This pass runs on the JOINT graph produced by AOT Autograd, prior to
+# partitioning, because we want to be able to make changes that affect
+# our partitioning decisions (in particular, we want to avoid having to
+# be able to save floats across the partition, and passes that change what
```
I can't read /understand this part.
" and passes that change what device compute happen on need to happen before partitioning, but after this
pass"

bobrenjc93 replied:
1) This pass eliminates floats by replacing TensorScalar operations with TensorTensor operations and then DCEing the item calls (eliminating the floats)
2) There is a separate pass that changes what device compute happens on that needs to be done before partitioning but after this pass.

Therefore, due to these two constraints, we need to run this pass on the joint graph. I'll see if I can update the comment to be more understandable.

bobrenjc93 committed:
Update on "Tensorify compute on Python scalars"


Signed-off-by: Bob Ren <bobrenfb.com>

Comandeered from https://github.com/pytorch/pytorch/pull/130228 as I'm helping ezyang w/ shipping dynamic float arguments in PT2. This starts with supporting torch.ops.aten.mul. I'll stack on top support for other operators in subsequent PRs to keep this scoped to the mechanics of the fx pass.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

ezyang commented on torch/fx/proxy.py:
```diff
@@ -529,6 +529,24 @@ def find_tracer(a):
+        kwargs = kwargs if kwargs else {}
+
+        proxy = super().__torch_function__(orig_method, types, args, kwargs)
+        proxy.node.meta["val"] = orig_method(*[a.node.meta["val"] if isinstance(a, Proxy) else a for a in args], **kwargs)
```
It might be safer to explicitly enable the fake tensor mode here before doing the prop. This saves you in a situation where MetaProxy is representing a non-Tensor argument (which can happen with plain Proxy usage, although it doesn't happen for your code, I think), and consequently you end up doing a function call without any fake tensor inputs, and then you end up allocating a real tensor. This is probably most conveniently done by explicitly associating a fake mode with the proxy.

ezyang commented on torch/fx/passes/_tensorify_python_scalars.py:
```diff
@@ -0,0 +1,270 @@
+            node = graph.call_function(
+                torch.ops.aten.scalar_tensor.default, (c,), {"dtype": dtype}
+            )
+            node.meta["val"] = torch.ops.aten.scalar_tensor.default(c, dtype=dtype)
```
This forces a precondition that the fake tensor mode must have already been enabled on entry of this function. It seems better hygiene if we explicitly `with fake_mode` before doing this

ezyang commented on torch/fx/passes/_tensorify_python_scalars.py:
```diff
@@ -0,0 +1,270 @@
+            expr_to_tensor_proxy[expr] = MetaProxy(
+                node,
+                tracer=tracer,
+            )
```
Why don't you just use MetaProxy here? (so directly torch.ops.aten.scalar_tensor.default(expr_to_sym_proxy[expr].node))

ezyang commented on torch/fx/passes/_tensorify_python_scalars.py:
```diff
@@ -0,0 +1,270 @@
+                                "val"
+                            ] = torch.ops.prims.convert_element_type.default(
+                                a.meta["val"], compute_dtype
+                            )
```
MetaProxy use site here too?

ezyang commented on torch/fx/passes/_tensorify_python_scalars.py:
```diff
@@ -0,0 +1,270 @@
+                    )
+                    replacement_node.meta["val"] = torch.ops.aten.mul.Tensor(
+                        *[a.meta["val"] for a in args]
+                    )
```
and here

ezyang commented on torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py:
```diff
@@ -179,6 +181,10 @@ def aot_dispatch_base(
+            fake_mode = detect_fake_mode()
+            if fake_mode is not None:
+                assert isinstance(fw_module, GraphModule)
+                tensorify_python_scalars(fw_module, fake_mode.shape_env, fake_mode)
```
hmm, I hope it's ok to mutate fw_module here...

ezyang reviewed (approved):
Nice work!

bobrenjc93 commented:
Comments address in https://github.com/pytorch/pytorch/pull/137588

@pytorchbot merge
